src/sg/aca_security_group_manager.cpp

```diff
@@ -0,0 +1,273 @@
+		return update_security_group_rule(port, sg, sg_rule); 
+	}
+
+	if (set_remote_group(sg_rule) != EXIT_SUCCESS) {
+		return EXIT_FAILURE;
+	}
+
+	sg_rule.set_cookie(0);
+	sg_ovs.create_port_security_group_rule(port, sg_rule);
+	sg.add_security_group_rule(&sg_rule); 
+
+	return EXIT_SUCCESS;
+}
+
+int Aca_Security_Group_Manager::update_security_group_rule(Aca_Port &port,
+													  Aca_Security_Group &sg,
+                                                      Aca_Security_Group_Rule &sg_rule) 
+{	
+	Aca_Security_Group_Ovs &sg_ovs = Aca_Security_Group_Ovs::get_instance();
+	string rule_id;
+	Aca_Security_Group_Rule *aca_sg_rule;
+	rule_id = sg_rule.get_id();
+	
+	aca_sg_rule = sg.get_security_group_rule(rule_id);
+	if (aca_sg_rule == NULL) {
+		ACA_LOG_ERROR("Can not find security group rule by id:%s", rule_id.data());
+		return EXIT_FAILURE; 
+	}
+
+	if (set_remote_group(sg_rule) != EXIT_SUCCESS) {
+		return EXIT_FAILURE;
+	}
+
+	sg_rule.set_cookie(aca_sg_rule->get_cookie() + 1);
+	sg_ovs.update_port_security_group_rule(port, sg_rule, *aca_sg_rule);
+	sg.update_security_group_rule(&sg_rule); 
+
+	return EXIT_SUCCESS;
+}
+
+int Aca_Security_Group_Manager::delete_security_group_rule(Aca_Port &port,
+													  Aca_Security_Group &sg,
+                                                      Aca_Security_Group_Rule &sg_rule) 
+{
+	Aca_Security_Group_Ovs &sg_ovs = Aca_Security_Group_Ovs::get_instance();
+	string rule_id;
+	Aca_Security_Group_Rule *aca_sg_rule;
+	rule_id = sg_rule.get_id();
+
+	aca_sg_rule = sg.get_security_group_rule(rule_id);
+	if (aca_sg_rule == NULL) {
+		ACA_LOG_ERROR("Can not find security group rule by id:%s", rule_id.data());
+		return EXIT_FAILURE; 
+	}
+	
+	if (set_remote_group(sg_rule) != EXIT_SUCCESS) {
+		return EXIT_FAILURE;
+	}
+
+	sg_ovs.delete_port_security_group_rule(port, *aca_sg_rule);
+	sg.delete_security_group_rule(sg_rule.get_id()); 
+
+	return EXIT_SUCCESS;
+}
+
+int Aca_Security_Group_Manager::create_security_group(Aca_Port &input_port,
+                                                 Aca_Security_Group &input_sg) 
+ {
+ 	Aca_Security_Group_Ovs &sg_ovs = Aca_Security_Group_Ovs::get_instance();
+ 	map<string, Aca_Port *>::iterator piter;
+	map<string, Aca_Security_Group *>::iterator siter;
+	Aca_Security_Group *aca_sg;
+	Aca_Port * aca_port;
+	string port_id = input_port.get_id();
+	string sg_id = input_sg.get_id();
+
+	//TODO: maybe multi threads access this map
+	piter = this->ports.find(port_id);
+    if (piter == this->ports.end()) {
+    	aca_port = new Aca_Port(input_port);
+    	this->ports[port_id] = aca_port;
+        sg_ovs.init_port_flows(*aca_port);
+    } else {
+		aca_port = piter->second;
+    }
+
+    siter = this->security_groups.find(sg_id);
+    if (siter == this->security_groups.end()) {
+        aca_sg = new Aca_Security_Group(input_sg);
+        this->security_groups[sg_id] = aca_sg;
+    } else {
+		aca_sg = siter->second;
+    }
+
+    aca_sg->add_port_id(port_id);
```

Comment:
note that we also keep track of ports inside vlan manager for other usage. see ACA_Vlan_Manager::get_instance().add_ovs_port. I think it is okay for security group manager to track ports seperately for now but can consider merging in the future.